grr/server/grr_response_server/gui/api_plugins/metadata_test.py

```diff
@@ -0,0 +1,382 @@
+  protobuf = tests_pb2.MetadataTypesHierarchyCyclic
+  rdf_deps = [
+    MetadataTypesHierarchyRoot,
+    "MetadataTypesHierarchyLeaf",
+  ]
+
+
+class MetadataTypesHierarchyLeaf(rdf_structs.RDFProtoStruct):
+  protobuf = tests_pb2.MetadataTypesHierarchyLeaf
+
+
+class DummyApiCallRouter(api_call_router.ApiCallRouter):
+  """Dummy ApiCallRouter implementation used for Metadata testing."""
+
+  @api_call_router.ArgsType(MetadataSimpleMessage)
+  @api_call_router.Http("GET", "/metadata_test/method1/<metadata_id>")
+  @api_call_router.Http("HEAD", "/metadata_test/method1/<metadata_id>")
+  @api_call_router.Http("POST", "/metadata_test/method1/<metadata_id>")
+  def Method1WithArgsType(self, args, token=None):
+    """Method 1 description."""
+
+  @api_call_router.ResultType(MetadataSimpleMessage)
+  @api_call_router.Http("GET", "/metadata_test/method2")
+  @api_call_router.Http("HEAD", "/metadata_test/method2")
+  @api_call_router.Http("POST", "/metadata_test/method2")
+  def Method2WithResultType(self, args, token=None):
+    """Method 2 description."""
+
+  @api_call_router.ArgsType(MetadataPrimitiveTypesMessage)
+  @api_call_router.ResultBinaryStream()
+  @api_call_router.Http("GET", "/metadata_test/method3")
+  def Method3PrimitiveTypes(self, args, token=None):
+    """Method 3 description."""
+
+  @api_call_router.ArgsType(MetadataRepeatedFieldMessage)
+  @api_call_router.Http("GET", "/metadata_test/method4")
+  @api_call_router.Http("POST", "/metadata_test/method4")
+  def Method4RepeatedField(self, args, token=None):
+    """Method 4 description."""
+
+  @api_call_router.ArgsType(MetadataEnumFieldMessage)
+  @api_call_router.Http("GET", "/metadata_test/method5")
+  @api_call_router.Http("POST", "/metadata_test/method5")
+  def Method5EnumField(self, args, token=None):
+    """Method 5 description."""
+
+  @api_call_router.ArgsType(MetadataTypesHierarchyRoot)
+  @api_call_router.ResultType(MetadataTypesHierarchyLeaf)
+  @api_call_router.Http("GET", "/metadata_test/method6")
+  def Method6TypeReferences(self, args, token=None):
+    """Method 6 description."""
+
+
+class ApiGetOpenApiDescriptionHandlerTest(api_test_lib.ApiCallHandlerTest):
+  """Test for ApiGetOpenApiDescriptionHandler."""
+  def setUp(self):
+    super(ApiGetOpenApiDescriptionHandlerTest, self).setUp()
+    self.router = DummyApiCallRouter()
+    self.router_methods = self.router.__class__.GetAnnotatedMethods()
+
+    self.handler = metadata_plugin.ApiGetOpenApiDescriptionHandler(self.router)
+
+    result = self.handler.Handle(None, token=self.token)
+    self.openapi_desc = result.openapi_description
+    self.openapi_desc_dict = json.loads(self.openapi_desc)
+
+  @skip.If(
+    "openapi-spec-validator" not in {p.key for p in pkg_resources.working_set},
+    "The openapi-spec-validator module used for validating the OpenAPI "
+    "specification is not installed."
+  )
+  def testRendersValidOpenApiDescription(self):
+    import openapi_spec_validator
```

Comment:
Please open a Github issue to make the openapi spec validator a GRR test package dependency and add a TODO here to move this import to the top as soon as this issue is resolved.

Reply:
I opened the issue and added the comment in this test file, as well as in the integration test. Thank you!

Reply:
Looks good!